### PR TITLE
feat(ui-licenseobligation): License obligation changes

### LINKED ISSUE #1050: Changes on License Obligations
There are a couple of proposed changes for the obligations at licenses:

* In the license detail view, drop the "Add Obligation pane"
* Update the Admin -> Obligations -> Add Obligation drop valid for projects (because org and project obligation shall be pulled from the project side only)
* In the edit mode a license there should be a license obligation relation UI pane there in order to maintain links between licenses and obligations. User should add obligations similar as they add releases to projects. This section could be the third between the attributes and the text
* In non edit more the linked obligations are just displayed in the list

### PR DESCRIPTION
**Pull request by smrutis1**
[//]: # (Copyright Bosch.IO GmbH 2020)
[//]: # (This program and the accompanying materials are made)
[//]: # (available under the terms of the Eclipse Public License 2.0)
[//]: # (which is available at https://www.eclipse.org/legal/epl-2.0/)
[//]: # (SPDX-License-Identifier: EPL-2.0)

> Please provide a summary of your changes here.

Please refer the mentioned issue.

> * Which issue is this pull request belonging to and how is it solving it? (#1050 )
> * Did you add or update any new dependencies that are required for your change? - No

Issue:  
closes #1050 

* Need to test the editing the license and adding obligation to it.

> Have you implemented any additional tests? - No

### Checklist
Must:
- [ ] All related issues are referenced in commit messages and in PR

Signed-off-by: Smruti Prakash Sahoo <smruti.sahoo@siemens.com>

### CODE REVIEW

**Review by JaideepPalit — CHANGES_REQUESTED** [review]
Reviewed the PR

**Comment by JaideepPalit** [comment]
Tested the PR. Observed some issues.

1. Update Button is coming in bottom of the table. Cancel button not there in Linked obligation Page
**Issue**
image
**Expected**
image

2. Update not working as expected. Update Fields in both the tab`License and Linked Obligation`  -> Goto linked obligation -> click on update.
Issue 1- License Tab changes not updated
Issue 2 - Adds all obligation to license instead of selected one.
Expected behavior - There should be one Update button to update entire License/

3. Linked Obligation Table in Edit Page
    - Add Obligation type in table. 
    - sort not working for all relevant fields . 
image

4. Linked Obligation table in Details Page
image
Is it possible to add/update `Applies to Development`, `Applies to Distribution` and `Further Properties`?
Kindly remove fields from code which are not used anymore. Maintain same table in Edit and Details Page

5. Maintain Similar table Structure and Functionality for Details/Edit and Add Obligation after excluding irrelevant columns
image

6. Select Obligation Dialog Box
    - Sorting Searching is Disabled
    - Add Obligation Type
    - Select all Obligation checkbox missing
    - Break work for very long obligation inorder to maintain proper structure of Dialog box
image

7. Able to add same obligation multiple times
image

8. Delete Button not working

File: scripts/migrations/042_remove_validForProject_from_Obligation.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+# functions
+# ---------------------------------------
+
+def removeFieldName(resultFile, qryResult, fieldToBeRemoved):
+    log = {}
+    log['totalCount'] = len(qryResult)
+    print 'Removing field name '+fieldToBeRemoved
+    log['Updated obligation fields '+fieldToBeRemoved] = []
+    for entity in qryResult:
+        del entity[''+fieldToBeRemoved+'']
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply scripts/migrations/042_remove_validForProject_from_Obligation.py]: Changed it

File: libraries/lib-datahandler/src/main/java/org/eclipse/sw360/datahandler/common/SW360Utils.java [hunk 2]
```diff
@@ -612,7 +612,7 @@ public static Integer parseStringToNumber(String input) {
         obligationAlreadyPresent.entrySet().stream().forEach(e -> obligationStatusMap.remove(e.getKey()));
 
         Map<String, ObligationStatusInfo> mapOfObligations = obligations.stream().filter(Objects::nonNull)
-                .filter(o -> isValidForProjectOptional || o.isValidForProject())
+                .filter(o -> isValidForProjectOptional)
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply libraries/lib-datahandler/src/main/java/org/eclipse/sw360/datahandler/common/SW360Utils.java]: removed

File: frontend/sw360-portlet/src/main/resources/META-INF/resources/html/licenses/includes/detailAddTodo.jspf [hunk 3]
```diff
@@ -153,7 +201,51 @@
...
+                datatable = undefined;
+            }
+            $('#licenseTodoTableDialog > tbody  > tr').each(function() {
+                isSelectedObl = $('td:eq(0)', this).find("input[type='checkbox']").is(":checked");
+                if(isSelectedObl){
+                    oblId = $('td:eq(0)', this).find("input[type='checkbox']").val();
+                    oblIds.push(oblId);
+                    $OlTitle = $('td:eq(1)', this);
+                    $OlText = $('td:eq(2)', this);
+                    row = "<tr><td>" + $OlTitle.text() + "</td><td>" + $OlText.text() + "</td><td class='content-middle'><svg class='action lexicon-icon' data-row-id='oblLinkRow${ob.id}' data-obligation-title='${ob.title}'><title><liferay-ui:message key='delete' /></title><use href='/o/org.eclipse.sw360.liferay-theme/images/clay/icons.svg#trash'/></svg></td></tr>"
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply frontend/sw360-portlet/src/main/resources/META-INF/resources/html/licenses/includes/detailAddTodo.jspf]: Done

File: frontend/sw360-portlet/src/main/resources/META-INF/resources/html/licenses/includes/detailAddTodo.jspf [hunk 4]
```diff
@@ -19,79 +19,127 @@
...
+                    </p>
+				</div>
+			    <div class="modal-footer">
+			        <button type="button" class="btn btn-light" data-dismiss="modal"><liferay-ui:message key="cancel" /></button>
+			        <button type="button" class="btn btn-danger"><liferay-ui:message key="delete.link" /></button>
+			    </div>
+			</div>
+		</div>
+	</div>
+</div>
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply frontend/sw360-portlet/src/main/resources/META-INF/resources/html/licenses/includes/detailAddTodo.jspf:44]: Have used `$dialog = dialog.confirm()`

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/licenses/LicensesPortlet.java [hunk 5]
```diff
@@ -141,8 +142,27 @@ private void prepareEditView(RenderRequest request, RenderResponse response) {
...
+                    .collect(Collectors.toList());
+            request.setAttribute(KEY_OBLIGATION_LIST, obligations);
+
+            final License license = client.getByID(id,user.getDepartment());
+            Set<String> licenseOblIds = license.getObligationDatabaseIds();
+            List<Obligation> obls = new ArrayList<Obligation>();
+
+            if (licenseOblIds != null) {
+                obls = CommonUtils.nullToEmptyList(client.getObligationsByIds(Lists.newArrayList(licenseOblIds)));
+            }
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/licenses/LicensesPortlet.java]: Done

File: frontend/sw360-portlet/src/main/resources/META-INF/resources/html/licenses/edit.jsp [hunk 6]
```diff
@@ -50,7 +50,23 @@
...
+        <div class="row">
+                <div class="col">
+                    <div class="tab-content">
+                        <core_rt:if test="${not addMode}">
+                            <div id="tab-AddTodo" class="tab-pane <core_rt:if test="${selectedTab == 'tab-AddTodo'}">active show</core_rt:if>">
+                                <%@include file="/html/licenses/includes/detailAddTodo.jspf" %>
+                            </div>
+                            <div id="tab-AddLicense" class="tab-pane <core_rt:if test="${selectedTab == 'tab-AddLicense'}">active show</core_rt:if>">
+                            <div class="row portlet-toolbar">
             <div class="col-auto">
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply frontend/sw360-portlet/src/main/resources/META-INF/resources/html/licenses/edit.jsp]: Done

**Comment by smrutis1** [comment]
Thanks for review and suggestions. Have addressed most of the issues. Kindly have a look and share the feedback(if any).

Regarding point 4.  Have removed `Applies to Development`, `Applies to Distribution`  from the UI, But have kept the code for now in the backend as need to test the license import functionality also after removing these.

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/licenses/LicensesPortlet.java [hunk 7]
```diff
@@ -317,6 +359,9 @@ private License updateLicenseFromRequest(License license, ActionRequest request)
         license.setGPLv2Compat(gpl2compatibility);
         license.setGPLv3Compat(gpl3compatibility);
         license.setChecked(checked);
+        String obligationIds = request.getParameter("obligations");
+        List<String> oblIds = obligationIds == null || obligationIds.equals("") ? Lists.newArrayList() : Arrays.asList(obligationIds.split(","));
```
Comment (JaideepPalit) [inline frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/licenses/LicensesPortlet.java:363]: Is it possible to use `CommonUtils.isNotNullEmptyOrWhitespace()` same for line 457

File: libraries/lib-datahandler/src/main/java/org/eclipse/sw360/datahandler/common/SW360Utils.java [hunk 8]
```diff
@@ -612,7 +612,6 @@ public static Integer parseStringToNumber(String input) {
         obligationAlreadyPresent.entrySet().stream().forEach(e -> obligationStatusMap.remove(e.getKey()));
 
         Map<String, ObligationStatusInfo> mapOfObligations = obligations.stream().filter(Objects::nonNull)
-                .filter(o -> isValidForProjectOptional || o.isValidForProject())
```
Comment (JaideepPalit) [inline libraries/lib-datahandler/src/main/java/org/eclipse/sw360/datahandler/common/SW360Utils.java:615]: Should also be removed from method signature.

**Review by JaideepPalit — CHANGES_REQUESTED** [review]
Code looks fine to me , expect below two comments.

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/licenses/LicensesPortlet.java (hunk 7 above)
Comment (smrutis1) [reply frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/licenses/LicensesPortlet.java]: Have used

File: libraries/lib-datahandler/src/main/java/org/eclipse/sw360/datahandler/common/SW360Utils.java (hunk 8 above)
Comment (smrutis1) [reply libraries/lib-datahandler/src/main/java/org/eclipse/sw360/datahandler/common/SW360Utils.java:615]: Removed it

**Comment by smrutis1** [comment]
Thanks for comments. Have made the changes

**Review by JaideepPalit — APPROVED** [review]
Code looks good to me. Needs retest.

**Comment by mcjaeger** [comment]
restested - adding obligation from creation before from license admin portlets works. obligations can be added to project

**Comment by mcjaeger** [comment]
@JaideepPalit please consider merging after (if) travis wen through

**Pull request merged**